File: ARSimpleClub/ARSimpleClub/ViewController.m
//
//  ViewController.m
//  ARSimpleClub
//
//  Created by HanGyo Jeong on 25/12/2018.
//  Copyright © 2018 HanGyoJeong. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <ARSCNViewDelegate>

@property (nonatomic, strong) IBOutlet ARSCNView *sceneView;
@property (nonatomic, strong) SCNScene *scene;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (nonatomic) BOOL planeFound;

@end

    
@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setupLabel];
    [self setupSceneView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self setupScene];
    [self startSession];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    // Pause the view's session
    [self.sceneView.session pause];
}

- (void)setupLabel{
    self.statusLabel.numberOfLines = 0;
    self.statusLabel.text = @"Please Find a Dance Floor";
}

- (void)setupSceneView{
    self.sceneView.delegate = self;
    self.sceneView.autoenablesDefaultLighting = YES;
}

- (void)setupScene{
    self.scene = [SCNScene new];
    self.sceneView.scene = self.scene;
}

- (void)startSession{
    ARWorldTrackingConfiguration *configuration = [ARWorldTrackingConfiguration new];
    configuration.planeDetection = ARPlaneDetectionHorizontal;
    configuration.planeDetection = ARWorldAlignmentGravityAndHeading;
    [self.sceneView.session runWithConfiguration:configuration];
}

- (void)showDiscoBallWithAnchor:(ARPlaneAnchor *)anchor onNode:(SCNNode *)node{
    SCNNode *plane = [self planeFromAnchor:anchor];
    SCNNode *discoBall = [self discoBall:plane];
    
    [plane addChildNode:discoBall];
    [node addChildNode:plane];
}

#pragma mark - ARSCNViewDelegate
/*
 Tells the delegate that a SceneKit node corresponding to a new AR anchor has been added to the scene
 */
- (void)renderer:(id<SCNSceneRenderer>)renderer didAddNode:(SCNNode *)node forAnchor:(ARAnchor *)anchor{
    if(self.planeFound == NO){
        if([anchor isKindOfClass:[ARPlaneAnchor class]]){
            dispatch_async(dispatch_get_main_queue(), ^{
                self.planeFound = YES;
                self.statusLabel.text = @"DANCEFLOOR FOUND!!!";
                
                ARPlaneAnchor *placeAnchor = (ARPlaneAnchor *)anchor;
                [self showDiscoBallWithAnchor:placeAnchor onNode:node];
            });
        }
    }
}

- (void)session:(ARSession *)session didFailWithError:(NSError *)error {
    // Present an error message to the user
    
}

- (void)sessionWasInterrupted:(ARSession *)session {
    // Inform the user that the session has been interrupted, for example, by presenting an overlay
    
}

- (void)sessionInterruptionEnded:(ARSession *)session {
    // Reset tracking and/or remove existing anchors if consistent tracking is required
    
}

#pragma marker - Node Builders
- (SCNNode *)planeFromAnchor:(ARPlaneAnchor *)anchor{
    SCNPlane *plane = [SCNPlane planeWithWidth:anchor.extent.x height:anchor.extent.z];
    plane.firstMaterial.diffuse.contents = [UIColor clearColor];
    
    SCNNode *planeNode = [SCNNode nodeWithGeometry:plane];
    planeNode.position = SCNVector3Make(anchor.center.x, 0, anchor.center.z);
    planeNode.transform = SCNMatrix4MakeRotation(-M_PI * 0.5, 1, 0, 0);
    
    return planeNode;
}

- (SCNNode *)discoBall:(SCNNode *)node{
    SCNSphere *sphere = [SCNSphere sphereWithRadius:0.1];
    sphere.firstMaterial.diffuse.contents = [UIImage imageNamed:@"disco"];
    
    SCNNode *sphereNode = [SCNNode nodeWithGeometry:sphere];
    //sphereNode.position = SCNVector3Make(0, 0, 3);
    sphereNode.position = node.position;
    return sphereNode;
}

@end

File: ARDrawingSimple/ARDrawingSimple/ViewController.swift
//
//  ViewController.swift
//  ARDrawingSimple
//
//  Created by HanGyo Jeong on 06/01/2019.
//  Copyright © 2019 HanGyoJeong. All rights reserved.
//

import UIKit
import SceneKit
import ARKit

class ViewController: UIViewController, ARSCNViewDelegate {

    @IBOutlet var sceneView: ARSCNView!
    
    var screenCenter: CGPoint?
    
    let session = ARSession();
    
    var virtualObjectManager: VirtualObjectManager!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupScene()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        // Create a session configuration
        let configuration = ARWorldTrackingConfiguration()

        configuration.planeDetection = .horizontal
        
        // Run the view's session
        sceneView.session.run(configuration, options: [.resetTracking, .removeExistingAnchors])
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        // Pause the view's session
        sceneView.session.pause()
    }

    // MARK: - ARSCNViewDelegate
    
/*
    // Override to create and configure nodes for anchors added to the view's session.
    func renderer(_ renderer: SCNSceneRenderer, nodeFor anchor: ARAnchor) -> SCNNode? {
        let node = SCNNode()
     
        return node
    }
*/
    
    func session(_ session: ARSession, didFailWithError error: Error) {
        // Present an error message to the user
        
    }
    
    func sessionWasInterrupted(_ session: ARSession) {
        // Inform the user that the session has been interrupted, for example, by presenting an overlay
        
    }
    
    func sessionInterruptionEnded(_ session: ARSession) {
        // Reset tracking and/or remove existing anchors if consistent tracking is required
        
    }
    
    func setupScene()
    {
        sceneView.setUp()
        sceneView.delegate = self
        sceneView.session = session
    }
    
    var focusSquare: 
    
    func updateFocusSquare()
    {
        guard let screenCenter = screenCenter else { return }
        
        DispatchQueue.main.async {
            var objectVisible = false
            for object in self.virtualObjectManager.virtualObjects{
                if self.sceneView.isNode(object, insideFrustumOf: self.sceneView.pointOfView!){
                    objectVisible = true
                    break
                }
            }
            
            if objectVisible{
                
            }else{
                
            }
        }
    }
    
    //pragma - ARSCNViewDelegate
    func renderer(_ renderer: SCNSceneRenderer, updateAtTime time: TimeInterval) {
        updateFocusSquare()
    }
}
